pkg/client/ocm/paginated_request.go

```diff
@@ -0,0 +1,46 @@
+package ocm
+
+import pkgerrors "github.com/pkg/errors"
+
+type pageItem[Data any] interface {
+	Each(func(value Data) bool)
+}
+
+type paginatedResponse[Items any] interface {
+	Size() int
+	Items() Items
+}
+
+type paginatedRequest[RequestType any, ResponseType any] interface {
+	Size(int) RequestType
+	Page(int) RequestType
+	Send() (ResponseType, error)
+}
+
+// fetchPages calls OCM API with size and page request parameters, allowing for
+// paged access to the data. Example request type: *amsv1.QuotaCostListRequest,
+// with *amsv1.QuotaCost as the data in the response list.
+// Iteration stops when all or maxPages pages are retrieved, or f returns false.
+func fetchPages[RQ paginatedRequest[RQ, RS], RS paginatedResponse[I], I pageItem[Data], Data any](
+	request paginatedRequest[RQ, RS], pageSize int, maxPages int, f func(Data) bool) error {
+
+	req := request.Size(pageSize)
+	for page := 1; page <= maxPages; page++ {
+		response, err := req.Page(page).Send()
+		if err != nil {
+			return pkgerrors.Wrapf(err, "error retrieving page %d", page)
+		}
+		keepGoing := true
+		response.Items().Each(func(data Data) bool {
+			keepGoing = f(data)
+			return keepGoing
+		})
+		if !keepGoing || response.Size() < pageSize {
+			break
+		}
+		if page == maxPages {
+			return pkgerrors.New("too many pages")
+		}
+	}
```

Comment:
I think I need to correct my suggestion. I think it is still correct **not to fail** when `maxPages` page has `pageSize` elements.

```suggestion
	for page := 1; page <= maxPages + 1; page++ {
		response, err := req.Page(page).Send()
		if err != nil {
			return pkgerrors.Wrapf(err, "error retrieving page %d", page)
		}
		keepGoing := true
		response.Items().Each(func(data Data) bool {
			keepGoing = f(data)
			return keepGoing
		})
		if !keepGoing || response.Size() < pageSize {
			break
		}
		if page == maxPages + 1 &&  response.Size() > 0 {
			return pkgerrors.New("too many pages")
		}
	}
```

Reply:
While `response.Size() == pageSize` the result is considered incomplete. The logic of `maxPages` is to query not more than `maxPages`. Therefore, the function doesn't fetch `maxPages+1` page in the hope that last page was the last one.
Anyway, the `maxPages` feature is there mostly to prevent endless fetching, so there's no big difference if we set it to 1000 or 1001 in the real usecase.
I refactored the function to better indicate the error condition.